File: lib/pacer/pipe/vertices_pipe.rb
module Pacer::Pipes

  # This exists because the Blueprints GraphVerticesPipe includes the graph itself in the path
  class VerticesPipe < AbstractPipe
    attr_reader :starts

    def setStarts(starts)
      @starts = starts.first
      self.iter = @starts.getVertices.iterator
    end

    def processNextStart
      iter.next
    end

    def reset
      self.iter = @starts.getVertices.iterator
    end

    private

    attr_accessor :iter
  end
end

File: ext/src/main/java/com/xnlogic/pacer/pipes/VerticesPipe.java
package com.xnlogic.pacer.pipes;

import java.util.Iterator;

import com.tinkerpop.blueprints.Graph;
import com.tinkerpop.blueprints.Vertex;
import com.tinkerpop.pipes.AbstractPipe;

public class VerticesPipe extends AbstractPipe<Graph, Vertex> {
    private Iterator<Vertex> iter;
    private Graph starts;
      
    public void setStarts(Iterator<Graph> starts) {
        // TODO: Error checking?
        this.starts = (Graph)starts.next();
        this.iter = this.starts.getVertices().iterator();
    }

    protected Vertex processNextStart() {
        return this.iter.next();
    }

    public void reset() {
        this.iter = this.starts.getVertices().iterator();
    }
}
